README.md
# dbr: Convenient database connections and queries from R

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active) ![CRAN](https://www.r-pkg.org/badges/version/dbr) [![Build Status](https://travis-ci.org/daroczig/dbr.svg?branch=master)](https://travis-ci.org/daroczig/dbr) [![Code Coverage](https://codecov.io/gh/daroczig/dbr/branch/master/graph/badge.svg)](https://codecov.io/gh/daroczig/dbr)

Vignette coming, until then, please check the talk presented at the useR! 2018 conference:

* [slides](http://bit.ly/user2018-dbr)
* [video](https://www.youtube.com/watch?v=z7x4UOHNguY)

## Setting up a config file for the database connections

To be able to connect to a database, the connection parameters are to be specified in a YAML file, for example for a SQLite database to be created in a temp file:

```yaml
sqlite:
  drv: !expr RSQLite::SQLite()
  dbname: !expr tempfile()
```

By default, `dbr` will look for a file named `db_config.yaml` in the current working directory, that can be override via the `dbr.db_config_path` global option, eg to the example config bundled in this package:

```r
options(dbr.db_config_path = system.file('example_db_config.yaml', package = 'dbr'))
```

A more complex example from the demo YAML file describing a MySQL connection to a database hosted by RStudio (with public username and password):

```yaml
shinydemo:
  drv: !expr RMySQL::MySQL()
  host: shiny-demo.csa7qlmguqrf.us-east-1.rds.amazonaws.com
  username: guest
  password: guest
  dbname: shinydemo
```

Note, that instead of simple strings, you can also specify KMS-encrypted passwords, other secrets and parameters as well, eg:

```yaml
redshift:
  host: !aws_kms |
    KMSencryptedciphertext...
  port: 5439
  dbname: dbname
  user: username
  drv: !expr RPostgreSQL::PostgreSQL()
  password: !aws_kms |
    KMSencryptedciphertext...
  s3_copy_bucket: !attr |-
    's3://openmail-model/temp'
  s3_copy_iam_role: !attr |-
    arn:aws:iam::accountid:role/redshift_role
```

## Querying databases

Once the connection parameters are loaded from a config file, making SQL queries are as easy as specifying the SQL statement and the name of the connection:

```r
db_query('show tables', 'shinydemo')
#> INFO [2019-01-06 01:06:18] Connecting to shinydemo
#> INFO [2019-01-06 01:06:19] Executing:**********
#> INFO [2019-01-06 01:06:19] show tables
#> INFO [2019-01-06 01:06:19] ********************
#> INFO [2019-01-06 01:06:19] Finished in 0.1336 secs returning 3 rows
#> INFO [2019-01-06 01:06:19] Closing connection to shinydemo
#>   Tables_in_shinydemo
#> 1                City
#> 2             Country
#> 3     CountryLanguage
```

For more advanced usage, eg caching database connections, check `?db_connect` and the above mentioned vignette.

## SQL templating

To reuse SQL chunks, you may list your SQL queries (or parts of it) in a structured YAML file, like in the bundled example config at [`example_sql_chunks.yaml`](https://github.com/daroczig/dbr/blob/master/inst/example_sql_chunks.yaml)

Use `sql_chunk_files` to list or update the currently used SQL template YAML file(s), eg via

```r
sql_chunk_files(system.file('example_sql_chunks.yaml', package = 'dbr'))
```

Then you may refer to any key in that definition by a string that consist of the keys in hierarchy separated by a dot, so looking at the below definition (part of [`example_sql_chunks.yaml`](https://github.com/daroczig/dbr/blob/master/inst/example_sql_chunks.yaml)):

```yaml
dbr:
  shinydemo:
    countries:
      count: SELECT COUNT(*) FROM Country
```

Getting the `count` key from for the `countries` item in `dbr`'s `shinydemo` section, you could do something like:

```r
sql_chunk('dbr.shinydemo.countries.count')
#> SELECT COUNT(*) FROM Country
```

And pass it right away to `db_query`:

```r
countries <- db_query(sql_chunk('dbr.shinydemo.countries.count'), 'shinydemo')
#> INFO [2019-01-06 01:33:33] Connecting to shinydemo
#> INFO [2019-01-06 01:33:34] Executing:**********
#> INFO [2019-01-06 01:33:34] SELECT COUNT(*) FROM Country
#> INFO [2019-01-06 01:33:34] ********************
#> INFO [2019-01-06 01:33:34] Finished in 0.1291 secs returning 1 rows
#> INFO [2019-01-06 01:33:34] Closing connection to shinydemo
```

SQL chunks can be also defined in files outside of the YAML with the `sql` file extensions, and referenced with the `!include` tag in the YAML file, eg:

```yaml
dbr:
  shinydemo:
    countries:
      europe: !include europe.sql
```

This will read the content of [`europe.sql`](https://github.com/daroczig/dbr/blob/master/inst/europe.sql) and make it available as `sql_chunk('dbr.shinydemo.countries.count')`.

Besides files, a folder with `sql` files can be also included -- in that case, the base filename (without the `sql` file extension) will become the key under the given key. For example, consider this YAML definition:

```yaml
cities: !include cities.sql
```

Will load all the files from the [`cities.sql`](https://github.com/daroczig/dbr/tree/master/inst/cities.sql) folder and make those available under `europe`, so resulting in an intermediate YAML as:

```
cities: !include cities.sql
  europe: |-
    SELECT Name
    FROM City
    WHERE CountryCode IN (
      {sql_chunk('dbr.shinydemo.countries.europe', indent_after_linebreak = 2)})
  europe_large: |-
    SELECT Name
    FROM City
    WHERE
      Population > 1000000 AND
      Name IN (
        {sql_chunk('dbr.shinydemo.cities.europe', indent_after_linebreak = 4)}))
```

If the key of a directory `!include` is `~!`, then the keys are made available in the parent node, so eg

```
cities:
  ~!: !include cities.sql
```

Would not actually create the `cities` key, but only the `europe` and `europe_large` keys in the root node.

As you can see from the above, the main power of this templating approach is that you can easily reuse SQL chunks, eg for the list of European countries in:

```r
cities <- db_query(sql_chunk('dbr.shinydemo.cities.europe'), 'shinydemo')
#> INFO [2019-01-06 01:32:02] Connecting to shinydemo
#> INFO [2019-01-06 01:32:02] Executing:**********
#> INFO [2019-01-06 01:32:02] SELECT Name
#> FROM City
#> WHERE CountryCode IN (
#>   SELECT Code
#>   FROM Country
#>   WHERE Continent = 'Europe')
#> INFO [2019-01-06 01:32:02] ********************
#> INFO [2019-01-06 01:32:02] Finished in 0.1225 secs returning 643 rows
#> INFO [2019-01-06 01:32:02] Closing connection to shinydemo
```

Where the `Country`-related subquery was specified in the `dbr.shinydemo.countries.europe` key as per:

```sql
SELECT Name
FROM City
WHERE CountryCode IN (
  {sql_chunk('dbr.shinydemo.countries.europe', indent_after_linebreak = 2)})
```

The `indent_after_linebreak` parameter is just for cosmetic updates in the query to align `FROM` and `WHERE` on the same character in the SQL statement.

Even more complex / nested example:

```sql
sql_chunk('dbr.shinydemo.cities.europe_large')
#> SELECT Name
#> FROM City
#> WHERE
#>   Population > 1000000 AND
#>   Name IN (
#>     SELECT Name
#>     FROM City
#>     WHERE CountryCode IN (
#>       SELECT Code
#>       FROM Country
#>       WHERE Continent = 'Europe')))
```


NEWS.md
# dbr (master branch)

* create `pkgdown` site for documentation
* better unit-test coverage
* test-phase of Mikata Project integration

# dbr 0.1.0.9004 (2020-06-08)

* add `aws` prefix for the AWS-related secrets handler(s)
* add function parameter for passing in temporarily SQL formatter function
* make `botor` dependency optional
* improve documentation and examples
* add support for reading config values from AWS System Manager's Parameter Store
* add support for SQL chunks to be defined in files and folders outside of the YAML definitions

# dbr 0.1.0.9003 (2019-01-15)

* add support for `gzip` when inserting records into Redshift

# dbr 0.1.0.9002 (2019-01-14)

* add helper function to write JSON lines
* add support for passing database parameters not required for making the connection
* add support for inserting records into Redshift via S3 staging and `COPY FROM`

# dbr 0.1.0.9001 (2019-01-09)

* optionally return `data.table` or `tibble` instead of base `data.frame`
* improve documentation
* more unit tests

# dbr 0.1.0.9000 (2019-01-09)

* switch versioning to "in-development" schema
* drop `futile.logger` dependency for `logger`
* drop `AWR` dependencies for `botor`
* add functionality to write to tables
* improve documentation and examples on YAML definitions
* introduce support for SQL templating

# dbr 0.1 (2018-07-04)

Initial release after open-sourcing at System1.


inst/tests/testthat.R
library(testthat)
library(dbr)

test_check('dbr')


inst/tests/testthat/test-db.R
library(dbr)
library(testthat)

library(logger)
log_threshold_bak <- log_threshold()
log_threshold(FATAL)
log_threshold(FATAL, namespace = 'dbr')
example_layout <- layout_glue_generator(format = '{node}/{pid}/{namespace}/{fn} {time} {level}: {msg}')
log_layout(example_layout, namespace = 'dbr')

config_path_bak <- getOption('dbr.db_config_path')
options('dbr.db_config_path' = system.file('example_db_config.yaml', package = 'dbr'))

## #############################################################################

context('DB helpers')

con <- db_connect('sqlite')
test_that('connection', {
    expect_s4_class(con, 'SQLiteConnection')
})

test_that('static sql', {
    expect_equal(db_query('select 42', db = con)[[1]], 42)
    expect_equal(db_query('select 42', con)[[1]], 42)
    expect_equal(db_query('select 42', db = 'sqlite')[[1]], 42)
})
db_close(con)

test_that('cache', {
    con <- db_connect('sqlite', cache = TRUE)
    expect_equal(db_query('select 42', db = con)[[1]], 42)
    expect_equal(db_query('select 42', 'sqlite')[[1]], 42)
    con <- db_connect('sqlite', cache = FALSE)
    expect_equal(db_query('select 42', db = con)[[1]], 42)
    db_close(con)
})

test_that('refresh', {
    res <- db_query('select 42', db = 'sqlite')
    expect_equal(db_refresh(res)[[1]], 42)
    expect_true(attr(res, 'when') < attr(db_refresh(res), 'when'))
})

test_that('tibble', {
    expect_true(inherits(
        db_query('select 42', db = 'sqlite', output_format = 'tibble'),
        'tbl_df'
    ))
})

## #############################################################################

context('sql chunks')

bak_chunk_files <- sql_chunk_files()
sql_chunk_files(system.file('example_sql_chunks.yaml', package = 'dbr'))
test_that('static chunk works', {
    expect_equal(sql_chunk('dbr.unittests.static'), 'x')
})

test_that('variable substitution with fallback works', {
    expect_equal(sql_chunk('dbr.unittests.fallback'), 'x == 5')
    expect_equal(sql_chunk('dbr.unittests.fallback', x = 42), 'x == 42')
})

test_that('nested chunk works', {
    expect_equal(sql_chunk('dbr.unittests.nested'), 'xx == x')
    expect_equal(sql_chunk('dbr.unittests.nested2'), 'xxx == 42 | xx == x')
    expect_equal(sql_chunk('dbr.unittests.nested2', x = 6), 'xxx == 6 | xx == x')
})

test_that('global env overrides works', {

    t <- tempfile()

    cat('
      library(dbr)
      sql_chunk_files(system.file("example_sql_chunks.yaml", package = "dbr"))
      x <- 42
      sql_chunk("dbr.unittests.fallback")', file = t)
    expect_equal(
        system(paste('Rscript', t), intern = TRUE),
        'x == 42')

    cat('
      library(dbr)
      sql_chunk_files(system.file("example_sql_chunks.yaml", package = "dbr"))
      x <- 42
      sql_chunk("dbr.unittests.fallback", x = 0)', file = t)
    expect_equal(
        system(paste('Rscript', t), intern = TRUE),
        'x == 0')

    cat('
      library(dbr)
      sql_chunk_files(system.file("example_sql_chunks.yaml", package = "dbr"))
      x <- 42
      f <- function() sql_chunk("dbr.unittests.fallback", x = 0)
      f()', file = t)
    expect_equal(
        system(paste('Rscript', t), intern = TRUE),
        'x == 0')

    unlink(t)

    expect_equal(sql_chunk('dbr.unittests.fallback'), 'x == 5')
})

## #############################################################################

context('sql formatter')

formatter_bak <- getOption('dbr.sql_formatter')
library(glue)
options('dbr.sql_formatter' = glue)
.x <- 2

con <- db_connect('sqlite')
test_that('glueing', {
    expect_equal(db_query('SELECT 42', db = con)[[1]], 42)
    expect_equal(db_query('SELECT {42}', db = con)[[1]], 42)
    expect_equal(db_query('SELECT {40 + 2}', db = con)[[1]], 42)
    expect_equal(db_query('SELECT {40 + x}', x = 2, db = con)[[1]], 42)
    expect_equal(db_query('SELECT {y + x}', y = 40 , x = 2, db = con)[[1]], 42)
    expect_equal(db_query('SELECT {40 + .x}', db = con)[[1]], 42)
})
db_close(con)

## #############################################################################

rm(.x)
options('dbr.db_config_path' = config_path_bak)
options('dbr.sql_formatter' = formatter_bak)
sql_chunk_files(bak_chunk_files, add = FALSE)
log_threshold(log_threshold_bak)


inst/tests/testthat/test-checkmate.R
library(dbr)
library(testthat)

context('checkmate')

test_that('check_attr', {
    expect_true(check_attr(mtcars, 'names'))
    expect_equal(check_attr(mtcars, 'foobar'), "'mtcars' does not have the foobar attribute")
    expect_error(assert_attr(mtcars, 'foobar'))
})

test_that('assert_botor_available', {
    if ('botor' %in% rownames(installed.packages())) {
        expect_error(assert_botor_available(), NA)
    } else {
        expect_error(assert_botor_available())
    }
})


R/zzz.R
dbs <- new.env(hash = TRUE, parent = emptyenv(), size = 29L)

.onLoad <- function(libname, pkgname) {

    ## use sprintf for the logger formatter not to depend on glue
    logger::log_formatter(logger::formatter_sprintf, namespace = pkgname)

    ## path to the default DB config YAML file
    options('dbr.db_config_path' = getOption(
        x = 'dbr.db_config_path',
        function() file.path(getwd(), 'db_config.yaml')))

    ## default SQL formatter does not do anything
    options('dbr.sql_formatter' = getOption(
        x = 'dbr.sql_formatter',
        default = identity))

    ## by default, return data as standard data.frame
    ## as data.table or tibble might not be available / preferred
    options('dbr.output_format' = getOption(
        x = 'dbr.output_format',
        default = 'data.frame'))

}



R/chunks.R
chunkfiles <- NULL


#' List or update the list of SQL chunk files to be used in \code{\link{sql_chunk}}
#' @param file path
#' @param add by default, the new \code{file} will be added to the list of active SQL chunk files, but when set to \code{FALSE}, it will override the old list instead of appending
#' @return character vector of the active SQL chunk files (invisibly after update)
#' @export
#' @importFrom utils assignInMyNamespace
sql_chunk_files <- function(file, add = TRUE) {

    if (!missing(file)) {
        if (add == TRUE) {
            newchunkfiles <- c(chunkfiles, file)
        } else {
            newchunkfiles <- file
        }
        assignInMyNamespace('chunkfiles', newchunkfiles)
        return(invisible(newchunkfiles))
    }

    unique(chunkfiles)

}


#' Look up all SQL chunks from YAML definitions
#' @return list
#' @export
sql_chunks <- function() {

    ## path where looking for SQL chunk files
    paths <- dirname(sql_chunk_files())

    ## parse config file(s)
    chunk <- unlist(lapply(sql_chunk_files(), function(chunkfile) {
        if (!file.exists(chunkfile)) {
            log_warn('%s SQL chunk file not found', chunkfile)
        } else {
            yaml.load_file(
                chunkfile,
                handlers = list('include' = function(x) structure(x, class = 'include')))
        }
    }), recursive = FALSE)

    ## read SQL bits from files
    chunk <- rapply(chunk, function(chunk) {

        ## normal chunk, exit early
        if (!inherits(chunk, 'include')) {
            return(chunk)
        }

        files <- list.files(paths, pattern = '*[\\.-]sql$', full.names = TRUE)
        files <- files[basename(files) == chunk]

        if (length(files) < 0) {
            stop('SQL chunk file not found at ',
                 paste(file.path(paths, chunk), collapse = '; '))
        }

        if (length(files) > 1) {
            stop('Multiple SQL chunk files found at ',
                 paste(file.path(paths, chunk), collapse = ' and '))
        }

        ## chunk defined in a file
        if (isFALSE(file.info(files)$isdir)) {
            log_trace('Found a file reference for %s at %s', chunk, files)
            return(paste(readLines(files, warn = FALSE), collapse = '\n'))
        }

        ## chunks defined in a folder
        log_trace('Found a folder reference for %s at %s', chunk, files)
        files <- list.files(files, full.names = TRUE)
        setNames(lapply(files, function(file) {
            paste(readLines(file, warn = FALSE), collapse = '\n')
        }), sub('.sql$', '', basename(files)))

    }, how = 'replace')

    ## handle special meaning of `~!` index (moving a level up)
    list_remove_intermediate_level_by_name(chunk, '~!')

}


#' Look up common SQL chunks from YAML definitions to be reused in SQL queries
#'
#' For more details and examples, please see the package \code{README.md}.
#' @param key optional key defined in \code{\link{sql_chunk_files}} to filter for
#' @param ... passed to \code{glue} for string interpolation
#' @param indent_after_linebreak integer for extra indent
#' @return string
#' @export
#' @importFrom glue glue
#' @examples \dontrun{
#' sql_chunk_files(system.file('example_sql_chunks.yaml', package = 'dbr'))
#' sql_chunk('dbr.shinydemo.countries.count')
#'
#' ## pass it right away to a database
#' countries <- db_query(sql_chunk('dbr.shinydemo.countries.count'), 'shinydemo')
#'
#' ## example for a more complex query
#' cities <- db_query(sql_chunk('dbr.shinydemo.cities.europe'), 'shinydemo')
#' }
#' @importFrom logger log_trace log_warn %except%
sql_chunk <- function(key, ..., indent_after_linebreak = 0) {

    chunk <- sql_chunks()

    for (keyi in strsplit(key, '.', fixed = TRUE)[[1]]) {

        if (!hasName(chunk, keyi)) {
            stop(shQuote(keyi), ' from ', key, ' not found in the SQL chunk files')
        }

        ## get the SQL chunk
        chunk <- chunk[[keyi]]

    }

    ## string interpolation
    chunk <- do.call(glue, c(list(chunk, .trim = FALSE), list(...)))

    ## optional extra indent
    indent_spaces <- paste(rep(' ', indent_after_linebreak), collapse = '')
    gsub('\n', paste0('\n', indent_spaces), chunk)

}


#' Remove intermediate list by moving up the children by one level
#' @param l list
#' @param n list name of levels to be removed (but keeping children)
#' @return list
#' @keywords internal
#' @examples \dontrun{
#' l <- list(a = 1, b = list('removeme' = 2))
#' list_remove_intermediate_level_by_name(l, 'removeme')
#' }
list_remove_intermediate_level_by_name <- function(l, n) {
    if (!is.list(l)) return(l)
    if (length(l) == 1 && names(l) == n) {
        return(l[[1]])
    }
    for (i in which(names(l) == n)) {
        l <- c(l[-i], l[[i]])
    }
    return(lapply(l, list_remove_intermediate_level_by_name, n = n))
}


R/config.R
#' Load DB connection parameters
#'
#' Load DB connection parameters from a YAML file and optionally decrypt secrets via Amazon KMS or load values from Amazon System Manager's Parameter Store.
#'
#' The YAML file should include all connection parameters, including the Database Driver as an R object, see eg
#'
#' \preformatted{dbreference:
#'   host: localhost
#'   port: 3306
#'   drv: !expr RMySQL::MySQL()
#'   username: foo
#'   password: bar}
#'
#' KMS-encrypted secrets should be specified with the \code{aws_kms} type in the YAML file, such as:
#'
#' \preformatted{dbreference:
#'   ...
#'   password: !aws_kms |
#'     ciphertext}
#'
#' AWS Parameter Store values can be loaded by specifying the \code{aws_parameter} type in the YAML file, such as:
#'
#' \preformatted{dbreference:
#'   ...
#'   password: !aws_parameter |
#'     /path/to/value}
#'
#' Fields that should not be passed to \code{drv} (eg extra params not used for making the DB connection) should be specified with the \code{attr} type that will be added as attributes to the returned list.
#' @param db database name reference
#' @param db_config_path file path specified by the \code{dbr.db_config_path} global option as a string or function returning a string, which defaults to the \code{db_config.yaml} file in the current working directory
#' @return list of database parameters (eg \code{hostname}, \code{port}, \code{username}, \code{password} etc)
#' @export
#' @note You need to install the \code{botor} package and also have access to the related KMS key to be able to decrypt cipher-text.
#' @importFrom memoise memoise
#' @importFrom utils hasName
#' @importFrom yaml yaml.load_file
#' @importFrom logger log_debug
db_config <- memoise(function(db, db_config_path = getOption('dbr.db_config_path')) {

    if (is.function(db_config_path)) {
        db_config_path <- db_config_path()
    }

    if (!file.exists(db_config_path)) {
        stop(paste('DB config file not found at', db_config_path))
    }

    withclass <- function(class) {
        force(class)
        function(x) structure(x, class = class)
    }

    ## parse config file
    params <- yaml.load_file(
        db_config_path,
        ## keep classes
        handlers = list(
            'attr'          = withclass('attr'),
            ## legacy, use aws_kms instead, remove this in CRAN version
            'kms'           = withclass('aws_kms'),
            'aws_kms'       = withclass('aws_kms'),
            'aws_kms_file'  = withclass('aws_kms_file'),
            'aws_parameter' = withclass('aws_parameter')),
        eval.expr = TRUE)

    hasName(params, db) || stop('Database ', db, ' not found, check ', db_config_path)

    log_debug('Looking up config for %s', db)
    params <- params[[db]]

    ## check if we need botor package
    if (any(grepl('aws_', rapply(params, class, how = 'list')))) {
        if (!requireNamespace('botor', quietly = TRUE)) {
            stop('Please install the "botor" package to be able to use the AWS-specific configs.')
        }
    }

    ## hit KMS with each base64-encoded cipher-text (if any) and decrypt
    params <- rapply(params, function(param) {
        switch(
            class(param),

            ## decrypt base64-encoded ciphertext via Amazon KMS
            'aws_kms' = botor::kms_decrypt(param),

            ## decrypt file via a data encryption key and Amazon KMS
            'aws_kms_file' = {

                ## decrypt to tempfile
                t <- tempfile()
                on.exit(unlink(t))
                botor::kms_decrypt_file(param, return = t)

                ## load R object then cleanup
                readRDS(t)

            },

            ## get value from Amazon Systems Manager's Parameter Store
            'aws_parameter' = botor::ssm_get_parameter(param),

            ## default (no transformation)
            param)},
        how = 'replace')

    ## move attr list values from list to attributes
    attributes <- params[sapply(params, class) == 'attr']
    params <- params[setdiff(names(params), names(attributes))]
    for (attribute in names(attributes)) {
        attr(params, attribute) <- attributes[[attribute]]
    }

    params

})


#' Invalidates the cached secret storage
#' @export
#' @importFrom memoise forget
#' @importFrom logger log_info
db_config_invalidate_cache <- function() {
    log_info('Invalidating cache on already loaded DB config(s)')
    invisible(forget(db_config))
}


#' Encrypt a secret via KMS and prints the base64-encoded cipher-text to the console
#' @param secret string of the actual secret to be encrypted/stored
#' @param key Amazon KMS key to be used for the encryption
#' @return base64-encoded cipher-text
#' @export
#' @note You need to have access to the related KMS key to be able to decrypt cipher-text
#' @examples \dontrun{
#' encrypt_secret('secret sentence I want to store in the YAML file')
#' }
db_config_encrypt_secret <- function(secret, key) {
    assert_botor_available()
    structure(botor::kms_encrypt(key = key, text = secret), class = 'base64')
}

#' @export
print.base64 <- function(x, ...) {
    cat(sapply(
        split(strsplit(x, '')[[1]], rep(1:ceiling(nchar(x) / 76), each = 76, length.out = nchar(x))),
        paste, collapse = ''), sep = '\n')
}


R/redshift.R
#' Writes each row of a data frame as JSON into a file separated by line breaks as per \url{http://jsonlines.org}
#' @param df data frame
#' @param file path
#' @export
#' @examples \dontrun{
#' t <- tempfile()
#' write_jsonlines(mtcars, t)
#' }
#' @importFrom checkmate assert_data_frame assert_directory_exists
#' @importFrom logger fail_on_missing_package
write_jsonlines <- function(df, file = tempfile()) {
    ## TODO move to botor package?
    assert_data_frame(df)
    assert_directory_exists(dirname(file))
    fail_on_missing_package('jsonlite')
    jsonlite::stream_out(df, file(file), verbose = FALSE)
}

#' Checks if a database reference or connection is Redshift or not
#' @param db database reference by name or object
#' @return boolean
#' @keywords internal
is.redshift <- function(db) {
    config <- db_config(ifelse(is.object(db), attr(db, 'db'), db))
    class(config$drv) =="PostgreSQLDriver" && config$port == 5439
}

#' Dumps a data frame to disk, copies to S3 and runs COPY FROM on Redshift
#'
#' Note that the related database's YAML config should include \code{s3_copy_bucket} and \code{s3_copy_iam_role} fields with \code{attr} type pointing to a staging S3 bucket (to which the current node has write access, and the Redshift IAM Role has read access) and the full ARN of the Redshift IAM Role.
#' @param df \code{data.frame}
#' @param table Redshift schema and table name (separated by a dot)
#' @inheritParams db_query
#' @export
redshift_insert_via_copy_from_s3 <- function(df, table, db) {

    assert_botor_available()

    ## TODO split JSON into smaller chunks and provide a Manifest file
    ##      to speed up operations and also to fix current limit of ~4Gb data

    ## extract database name if actual DB connection was passed
    dbname <- ifelse(is.object(db), attr(db, 'db'), db)

    ## load and test if required params set for DB
    config <- db_config(dbname)
    if (!all(c('s3_copy_bucket', 's3_copy_iam_role') %in% names(attributes(config)))) {
        stop('Need to specify s3_copy_bucket and s3_copy_iam_role in the database config YAML')
    }

    ## dump data frame to S3 as jsonlines
    s3 <- tempfile(tmpdir = attr(config, 's3_copy_bucket', exact = TRUE), fileext = '.json.gz')
    botor::s3_write(df, write_jsonlines, s3, compress = 'gzip')

    ## load data from S3 into Redshift
    iam_role <- attr(config, 's3_copy_iam_role', exact = TRUE)
    db_query(
        paste(
            "COPY", paste(table, collapse = '.'),
            "FROM", shQuote(s3, type = 'sh'),
            "iam_role", shQuote(iam_role, type = 'sh'),
            "JSON 'auto' GZIP"),
        db = db)

    ## clean up
    botor::s3_delete(s3)

}



R/db.R
#' Connect to a database using YAML configs
#' @param db name of database, must be present in the file specified by \code{getOption('dbr.db_config_path')}
#' @param cache optional caching of the connection. If set to \code{TRUE}, the connection will be cached in the background and an all future \code{db_connect} calls will simply return that (even if called automatically from eg \code{db_query}) until the end of the R session or when caching on the \code{db} is disabled in a future \code{db_connect} call with explic \code{cache = FALSE}. See the examples for more details.
#' @param ... extra parameters passed to the database driver, even ones overriding the default values loaded from the YAML config
#' @importFrom DBI dbConnect dbDriver
#' @export
#' @seealso \code{\link{db_close}} \code{\link{db_query}}
#' @examples \dontrun{
#' ## create new connection
#' optbak <- options()
#' options('dbr.db_config_path' = system.file('example_db_config.yaml', package = 'dbr'))
#' con <- db_connect('sqlite')
#' str(con)
#' db_query('SELECT 42', 'sqlite')
#'
#' ## reusing the connection
#' str(db_connect('sqlite', cache = TRUE))
#' str(db_connect('sqlite'))
#' str(db_connect('sqlite'))
#' ## kill cached connection
#' db_close(db_connect('sqlite', cache = FALSE))
#'
#' ## restore options
#' options(optbak)
#' }
db_connect <- function(db, cache, ...) {

    cache <- ifelse(missing(cache), 'default', cache)

    if (exists(db, dbs)) {
        if (cache != FALSE) {
            return(dbs[[db]])
        }
        else {
            ## reset cached connection
            db_close(dbs[[db]])
            rm(list = db, envir = dbs)
        }
    }

    params <- db_config(db)

    ## override defaults
    extraparams <- list(...)
    for (i in seq_len(length(extraparams))) {
        params[[names(extraparams)[i]]] <- extraparams[[i]]
    }
    extralog <- ifelse(
        length(extraparams) > 0,
        paste0(' [', paste(paste(names(extraparams), extraparams, sep = '='), collapse = ', '), ']'),
        '')

    log_info(paste('Connecting to', db, extralog))
    con <- structure(do.call(dbConnect, params), db = db, cached = cache)

    ## cache connection
    if (isTRUE(cache)) {
        dbs[[db]] <- con
    }

    con

}


#' Close a database connection
#' @param db database object returned by \code{\link{db_connect}}
#' @importFrom DBI dbDisconnect
#' @export
#' @seealso \code{\link{db_connect}}
#' @note To close a cached connection, call \code{db_close} on an object returned by \code{db_connect(..., cache = FALSE)}
db_close <- function(db) {
    assert_attr(db, 'db')
    if (!isTRUE(attr(db, 'cached'))) {
        log_info(paste('Closing connection to', attr(db, 'db')))
        dbDisconnect(db)
    }
}


#' Execute an SQL query in a database
#' @param sql string
#' @param db database reference by name or object
#' @param ... passed to \code{sql_formatter}
#' @param sql_formatter function to be applied on \code{sql} potentially with \code{...}, eg using \code{glue} for string interpolation
#' @param output_format preferred output format that defaults to \code{data.frame}, but could be also \code{data.table} or \code{tibble} as well if the related R package is installed
#' @return data.frame with query metadata
#' @export
#' @importFrom DBI dbGetQuery
#' @importFrom logger log_info skip_formatter
#' @importFrom checkmate assert_string assert_function
#' @seealso \code{\link{db_connect}} \code{\link{db_refresh}}
#' @examples \dontrun{
#' options('dbr.db_config_path' = system.file('example_db_config.yaml', package = 'dbr'))
#' db_query('SELECT 42', 'sqlite')
#' db_query('SELECT {40 + 2}', 'sqlite')
#' }
db_query <- function(sql, db, ...,
                     sql_formatter = getOption('dbr.sql_formatter'),
                     output_format = getOption('dbr.output_format')) {

    if (!is.object(db)) {
        db <- db_connect(db)
        on.exit({
          db_close(db)
        })
    }

    assert_attr(db, 'db')
    assert_string(sql)
    assert_function(sql_formatter)

    sql <- do.call(sql_formatter, c(list(sql), list(...)), envir = parent.frame())

    log_info('Executing:**********')
    log_info(skip_formatter(sql))
    log_info('********************')

    start <- Sys.time()
    result_set <- dbGetQuery(db, sql)
    time_to_exec <- Sys.time() - start

    log_info('Finished in %s returning %s rows', format(time_to_exec, digits = 4), nrow(result_set))

    attr(result_set, 'when') <- start
    attr(result_set, 'db') <- attr(db, 'db')
    attr(result_set, 'time_to_exec') <- time_to_exec
    attr(result_set, 'statement') <-  sql

    ## convert to proffered output format
    if (output_format != 'data.frame') {
        result_set <- switch(
            output_format,
            'data.table' = data.table::setDT(result_set),
            'tibble' = tibble::as_tibble(result_set),
            stop('Unsupported output_format -- please use data.frame, data.table or tibble.'))
    }

    result_set

}


#' Refresh SQL query
#' @param x object returned by \code{db_query}
#' @seealso \code{\link{db_query}}
#' @importFrom checkmate assert_data_frame
#' @export
db_refresh <- function(x) {
    assert_data_frame(x)
    assert_attr(x, 'db')
    assert_attr(x, 'statement')
    with(attributes(x), db_query(statement, db))
}



#' Insert rows into a database table
#' @param df data.frame
#' @param table character vector of an optional schema and table name
#' @inheritParams db_close
#' @param ... further parameters passed to \code{dbWriteTable}, eg to modify \code{row.names} or \code{append} (depends on driver)
#' @seealso \code{RMySQL::\link[RMySQL]{dbWriteTable}}, \code{RPostgreSQL::\link[RPostgreSQL]{dbReadTable-methods}}
#' @importFrom DBI dbWriteTable
#' @importFrom checkmate assert_character
#' @export
#' @seealso \code{\link{db_append}}
#' @examples \dontrun{
#' options('db_config_path' = system.file('example_db_config.yaml', package = 'dbr'))
#' db_insert(mtcars, 'mtcars', 'sqlite')
#' db_append(mtcars, c('dm', 'mtcars'), 'sqlite')
#' }
db_insert <- function(df, table, db, ...) {

    if (!is.object(db)) {
        db <- db_connect(db)
        on.exit({
          db_close(db)
        })
    }

    assert_data_frame(df)
    assert_character(table, min.len = 1)
    assert_attr(db, 'db')

    log_info('Inserting %s rows into %s', nrow(df), paste(table, collapse = "."))
    dbWriteTable(conn = db, name = table, value = df, ...)

}


#' Append rows into a database table
#'
#' This is a wrapper around \code{\link{db_insert}} with the default parameters set to append to a table.
#' @inheritParams db_insert
#' @export
db_append <- function(df, table, db, ...) {
    ## check if it's Redshift, as COPY FROM stdin doesn't work there
    if (is.redshift(db)) {
        redshift_insert_via_copy_from_s3(df = df, table = table, db = db)
    } else {
        ## otherwise do a COPY FROM stdin
        db_insert(df, table, db, overwrite = FALSE, append = TRUE, row.names = FALSE, ...)
    }
}
